### PR TITLE
fix extra nesting of folders when unzipping in place

### PR DESCRIPTION
**Pull request by hyi**
### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
Address issue #4152. To test it, when unzipping a file, no extra nesting of folders will result from unzipping.

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4549](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4549)

File: theme/templates/includes/navbar.html [hunk 1]
```diff
@@ -246,7 +246,12 @@ <h4>Notifications <span v-if="tasks.length">(${tasks.length})</span></h4>
                                         <small v-else>${taskMessages[task.name].status[task.status]}.
                                             <span v-if="task.payload && (task.status === 'delivered' || task.status === 'completed')">
                                             Click <a :href="task.payload">here</a> to refresh user interface.
-                                            </span></small>
+                                            </span>
+                                            <span v-else="task.payload.includes('FileOverrideException') && task.status === 'failed'">
+                                            $('#unzip-resp-message').text(${task.payload});
```
[the remark this answers is not in the document]
Comment (hyi) [reply theme/templates/includes/navbar.html]: @devincowan Yeah, that is the key problem. Thanks for pinpointing the issue spot-on.

**Review by devincowan — CHANGES_REQUESTED** [review]
this is cool! I never saw the modal to choose whether to overwrite existing folder

File: theme/templates/includes/navbar.html [hunk 2]
```diff
@@ -246,7 +246,11 @@ <h4>Notifications <span v-if="tasks.length">(${tasks.length})</span></h4>
                                         <small v-else>${taskMessages[task.name].status[task.status]}.
                                             <span v-if="task.payload && (task.status === 'delivered' || task.status === 'completed')">
                                             Click <a :href="task.payload">here</a> to refresh user interface.
-                                            </span></small>
+                                            </span>
+                                            <span v-else-if="task.payload.includes('FileOverrideException') && task.status === 'failed'">
```
[the remark this answers is not in the document]
Comment (hyi) [reply theme/templates/includes/navbar.html:250]: Hmm. It works for me in my local development environment. I'll rebuild my local development environment to see if I see any issues. Or even better I can probably deploy this branch to a test VM. I see @devincowan you have new-riverdev reserved and @pkdash have new-testservices reserved. By any chance, you don't need the test VM any more so I can grab it for deploying this branch? @devincowan @pkdash

[the remark this answers is not in the document]
Comment (devincowan) [reply theme/templates/includes/navbar.html:250]: @hyi you can use new-riverdev if you want
also, perhaps I should just rebuild my local and try it again...

[the remark this answers is not in the document]
Comment (devincowan) [reply theme/templates/includes/navbar.html:250]: @hyi I rebuilt and it's working now for me locally. Thanks

File: theme/static/js/resourceLandingAjax.js [hunk 3]
```diff
@@ -928,9 +928,12 @@ function unzip_irods_file_ajax_submit(res_id, zip_with_rel_path) {
         data: {
             res_id: res_id,
             zip_with_rel_path: zip_with_rel_path,
-            remove_original_zip: "false"
+            remove_original_zip: "false",
+            overwrite: overwrite
         },
         success: function (task) {
+            $('#res_id').val(res_id);
```
Comment (devincowan) [inline theme/static/js/resourceLandingAjax.js:935]: watch out for id clash
https://github.com/hydroshare/hydroshare/blob/develop/irods_browser_app/templates/irods_upload_add.html#L15

File: theme/templates/resource-landing-page/modals.html [hunk 4]
```diff
@@ -1490,6 +1490,32 @@ <h4 class="modal-title">Remove content type</h4>
...
+                    <p id="unzip-resp-message"></p>
+                    <p>
+                        Click <strong>Proceed</strong> to confirm to override files. Otherwise, click <strong>Cancel</strong>.
+                    </p>
+                    <input type="hidden" id="res_id" value=""/>
+                    <input type="hidden" id="zip_with_rel_path" value=""/>
+                </div>
+
+                <div class="modal-footer">
+                    <button type="button" class="btn btn-primary" data-dismiss="modal">Cancel</button>
```
Comment (devincowan) [inline theme/templates/resource-landing-page/modals.html:1511]: if you choose cancel, the mouse cursor gets stuck in "progress"

File: theme/static/js/resourceLandingAjax.js [hunk 5]
```diff
@@ -919,7 +919,7 @@ function zip_by_aggregation_file_ajax_submit(res_id, aggregationPath, zipFileNam
     });
 }
 
-function unzip_irods_file_ajax_submit(res_id, zip_with_rel_path) {
+function unzip_irods_file_ajax_submit(res_id, zip_with_rel_path, overwrite) {
     $("#fb-files-container, #fb-files-container").css("cursor", "progress");
```
Comment (devincowan) [inline theme/static/js/resourceLandingAjax.js:923]: need to reverse this in the case that someone selects "cancel" from the modal

File: hs_core/tests/api/rest/test_unzip.py [hunk 6]
```diff
@@ -53,26 +53,25 @@ def setUp(self):
...
 
+        # extra folder named as test should not exist
         list_url = "/hsapi/resource/%s/folders/test/" % self.pid
         response = self.client.get(list_url, data={})
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
-
-        unzip_url = "/hsapi/resource/%s/functions/unzip/test.zip/" % self.pid
-        response = self.client.post(unzip_url, data={})
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```
Comment (devincowan) [inline hs_core/tests/api/rest/test_unzip.py:69]: Is this actually the desired behavior?
I got the impression reading issue https://github.com/hydroshare/hydroshare/issues/4152 that we wanted to avoid the nested files like `test/test` but that we still want to retain the `test/` first directory
Maybe I missed something in conversation...

File: theme/static/js/hs_file_browser.js [hunk 7]
```diff
@@ -2322,6 +2322,28 @@ $(document).ready(function () {
...
+        function afterDoneRequest() {
+            $('#btn-unzip-file-override-proceed').parent().find(".btn[data-dismiss='modal']").removeClass("disabled");
+            $("#unzip-override-confirm-dialog").modal('hide');
+        }
+        function afterFailRequest() {
+            afterDoneRequest();
+            refreshFileBrowser();
+        }
+
+        $.when.apply($, calls).done(afterDoneRequest);
```
Comment (devincowan) [inline theme/static/js/hs_file_browser.js:2342]: I've only ever seen deferred methods to jQuery.when() called using method chaining as in
```
$.when.apply($, calls).done(afterDoneRequest).fail(afterFailRequest);
```
https://api.jquery.com/deferred.fail/
Does it actually work the way you've written it here? If so, that is good learning for me...I thought they had to be chained.

**Review by devincowan — CHANGES_REQUESTED** [review]
This is great Hong! A few questions...

File: theme/templates/includes/navbar.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (hyi) [reply theme/templates/includes/navbar.html:250]: @devincowan Great to hear. I don't need to deploy to new-testservices then now that it works for you locally.

File: hs_core/tests/api/rest/test_unzip.py (hunk 6 above)
Comment (hyi) [reply hs_core/tests/api/rest/test_unzip.py:69]: @devincowan The issue #4152 is about not creating an extra directory using the name of the zip file without the extension .zip, so for this test case, what it used to work is to put all unzipped files into an extra ```test``` directory, hence the test asserts an extra ```test``` directory exists. If you unzip the same zip file again, it will create an extra ```test-1```, so there would be no need to check for same files names for override, since files are unzipped into a unique folder which guarantees override will never happen. My work to address this issue is to remove the extra folder, in this test case, no ```test``` extra folder will be created, instead, files will be unzipped directly to the folder where the zip file resides. If the zip file contains folders, those folders will remain as part of unzipping process, but no extra ```test``` folder will be created to put unzipped files into. Hope this explanation makes sense, but let me know if it is still confusing.

File: theme/static/js/hs_file_browser.js (hunk 7 above)
Comment (hyi) [reply theme/static/js/hs_file_browser.js:2342]: Based on my testing, it works as expected. I tried not to refresh browser when the async call is done successfully since otherwise, the temporary folder where the unzipped files are put would show up in the file browser. However, when the async call failed, refreshFileBrowser will be called. It appears to work as expected since before I made this change, the temporary intermediate folder showed up in the file browser, and after I made the change, the temporary folder does not show up any more. It is good to know the chaining rule for deferred methods to JQuery.when(), but it appears not chaining also works. Just did a search in the file and found the chaining is not used in other places as well, e.g., [here](https://github.com/hydroshare/hydroshare/blob/develop/theme/static/js/hs_file_browser.js#L631-L642).

Comment (devincowan) [reply theme/static/js/hs_file_browser.js:2342]: nice! thanks for the learning

File: hs_core/tests/api/rest/test_unzip.py (hunk 6 above)
Comment (devincowan) [reply hs_core/tests/api/rest/test_unzip.py:69]: okay I didn't know that we wanted "files will be unzipped directly to the folder where the zip file resides"
So when unzipping in-place a file `/devin.zip` I thought we wanted it to result in `/devin/*contents*` and the issue was that it previously resulted in `/devin/devin/*contents*`

Okay I'm on board now! Thanks

File: hs_core/views/utils.py [hunk 8]
```diff
@@ -1027,74 +1027,71 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
...
-        else:
-            unzip_path = istorage.unzip(zip_with_full_path)
-            res_files = link_irods_folder_to_django(resource, istorage, unzip_path, auto_aggregate)
-            if resource.resource_type == 'CompositeResource':
-                # make the newly added files part of an aggregation if needed
-                for res_file in res_files:
-                    resource.add_file_to_aggregation(res_file)
+                else:
+                    istorage.delete(destination_file)
+        if not overwrite and override_tgt_paths:
```
Comment (sblack-usu) [inline hs_core/views/utils.py:1064]: Should this check happen before all the aggregation manipulations that come before?

File: hs_core/views/utils.py [hunk 9]
```diff
@@ -1027,74 +1027,71 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
...
+            if (istorage.exists(destination_file)):
+                if not overwrite:
+                    override_tgt_paths.append(destination_file.split('data/contents/')[1])
+                elif resource.resource_type == "CompositeResource":
+                    aggregation_object = resource.get_file_aggregation_object(
+                        destination_file)
+                    if aggregation_object:
+                        aggregation_object.logical_delete(user)
                     else:
+                        logger.error("No aggregation object found for " + destination_file)
```
Comment (sblack-usu) [inline hs_core/views/utils.py:1060]: This isn't an error, not all files have an aggregation associated with them.

File: hs_core/views/utils.py (hunk 8 above)
Comment (sblack-usu) [reply hs_core/views/utils.py:1064]: ok, I see that now on line 1052. The overwrite check gets lost in all of it. We can reduce some of the nesting if `override_tgt_paths` is always created.  Something like this:
```
for file in res_files:
  destination_file = _get_destination_filename(file.name, unzipped_foldername)
  if (istorage.exists(destination_file)):
    if not overwrite:
      override_tgt_paths.append(destination_file.split('data/contents/')[1])
if not overwrite and override_tgt_paths:
  message = 'move would overwrite {}'.format(', '.join(override_tgt_paths))
  raise FileOverrideException(message)
for override_tgt_path in override_tgt_paths:
  aggregation_object = resource.get_file_aggregation_object(destination_file)
  if aggregation_object:
    aggregation_object.logical_delete(user)
  else:
    istorage.delete(destination_file)
```

**Comment by hyi** [comment]
@devincowan @sblack-usu I have addressed all your comments. See if this PR looks good now. Thanks!

**Review by devincowan — APPROVED** [review]

**Review by devincowan — CHANGES_REQUESTED** [review]
All looks good, but waiting to approve pending a question about unzipping files directly into "root" of the resource. Sounds like Martin might want to discuss on our Monday call Aug1

**Comment by hyi** [comment]
@devincowan Sure, we can discuss on Monday's call, but want to note here that unzipping files directly into a temporary folder at the "root" of the resource is intentional to minimize potential conflict with other resource folders. Regardless whether unzipping succeeds or fails, this temporary folder will be deleted. Since unzipping is running as a celery task, the celery task will not be interrupted by network connection or any other client activities, so this temporary folder should be cleaned up after the action regardless whether it succeeds or fails.

**Comment by devincowan** [comment]
@hong thanks, yes that makes sense. @martinseul and I just weren't sure whether _unzipping files directly to the folder where the zip file resides_ was confirmed as the desired behavior. Sounds like we just want to discuss one more time...

File: hs_core/views/utils.py [hunk 10]
```diff
@@ -1027,74 +1027,73 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
...
+            if istorage.exists(destination_file):
+                override_tgt_paths.append(destination_file)
+
+        if not overwrite and override_tgt_paths:
+            override_simplified_paths = [tgt_path.split('data/contents/')[1] for tgt_path in override_tgt_paths]
+            raise FileOverrideException('move would overwrite {}'.format(', '.join(override_simplified_paths)))
+
+        for override_tgt_path in override_tgt_paths:
+            if (istorage.exists(override_tgt_path)):
+                if resource.resource_type == "CompositeResource":
```
Comment (sblack-usu) [inline hs_core/views/utils.py:1060]: I think we can remove this check for composite resource as well.  Everything will have been migrated to composite by Aug 1.

**Review by sblack-usu — APPROVED** [review]

**Review by devincowan — APPROVED** [review]
@hyi this looks great! Sorry that I held it up for more discussion... no issue, it was just my lack of background

**Comment by hyi** [comment]
@devincowan thanks for triggering more discussion about this issue. The discussion resulted in a new issue created about unzipping to folder #4714 which should cover all the user unzipping needs.
